Framework: use Redux for layoutFocus

sirbrillig opened the pull request:
This PR aims to replace the `layoutFocus` module (a Flux store) with `state.ui.layoutFocus` in the Redux state tree, along with the appropriate actions and selectors.

The old Flux store module will now call the new Redux actions and selectors for its methods. There is only one place in the code (that I could find) where a component observes the Flux store for changes (`Layout`), and that has been updated to use a selector instead.

The old methods, with their replacements (note that the actions will require a dispatcher):

- `layoutFocus.set( area )` -> `setLayoutFocus( area )` (action)
- `layoutFocus.setNext( area )` -> `setNextLayoutFocus( area )` (action)
- `layoutFocus.next()` -> `activateNextLayoutFocus()` (action)
- `layoutFocus.getCurrent()` -> `getCurrentLayoutFocus( state )`


### Testing

First, make sure that when visiting a my-sites page, clicking on the site selector ("switch site") still works as expected (switches the focus to `sites` and then back to `content` when done).

Second, when visiting a my-sites page, adding `?sp` in the querystring should focus the sites list.

The code is calling the `setLayoutFocus()` method, which, if you see the sites list active, means that it updated the Flux store. Then check the Redux store (eg: using Redux Devtools) and be sure that `state.ui.layoutFocus.current` has been changed to 'sites'.

Test live: https://calypso.live/?branch=update/use-redux-for-layout-focus

sirbrillig commented:
The tests in `client/my-sites/upgrades/domain-management/list/test/index.js` reach into the component to call functions and check state and refs, which cause issues when the component is wrapped in a `Provider`.

If the tests just used component input and output (props and markup), they would still pass. I'm working to convert them, which may end up being a separate PR to make sure the tests are still good.

sirbrillig commented:
Note to self: try passing the required actions/selected values down to some of these components as props rather than `connecting` ones that weren't connected before. That should take care of breaking tests.

sirbrillig commented:
Ping @mtias and @aduth for review.

aduth commented on client/state/action-types.js:
```diff
@@ -125,6 +125,9 @@ export const JETPACK_SYNC_START_ERROR = 'JETPACK_SYNC_START_ERROR';
 export const JETPACK_SYNC_STATUS_REQUEST = 'JETPACK_SYNC_STATUS_REQUEST';
 export const JETPACK_SYNC_STATUS_SUCCESS = 'JETPACK_SYNC_STATUS_SUCCESS';
 export const JETPACK_SYNC_STATUS_ERROR = 'JETPACK_SYNC_STATUS_ERROR';
+export const LAYOUT_SET_FOCUS = 'LAYOUT_SET_FOCUS';
```
I think these should be `LAYOUT_FOCUS_SET` and `LAYOUT_NEXT_FOCUS_SET` (the third seems correct)

aduth commented on client/state/ui/layout-focus/reducer.js:
```diff
@@ -0,0 +1,55 @@
+			}
+			// Temporary while replacing module with this reducer
+			layoutFocusModule.set( action.area );
+			return Object.assign( {}, state, { current: action.area, previous: state.current } );
```
Not sure your feelings on it (since it's a stage 1 feature), but I find the [object spread operator](https://github.com/sebmarkbage/ecmascript-rest-spread) to be a little nicer to read:

```js
return {
	...state,
	current: action.area,
	previous: state.current
};
```

aduth commented on client/state/ui/layout-focus/reducer.js:
```diff
@@ -0,0 +1,55 @@
+const initialState = { current: null, previous: null, next: null };
+
+export default function layoutFocus( state = initialState, action ) {
+	const validAreas = [ 'content', 'sidebar', 'sites', 'preview' ];
```
Should we move this to a top-of-module constant so we're not redeclaring it every time the reducer is called?

aduth commented on client/state/ui/layout-focus/reducer.js:
```diff
@@ -0,0 +1,55 @@
+
+export default function layoutFocus( state = initialState, action ) {
+	const validAreas = [ 'content', 'sidebar', 'sites', 'preview' ];
+	const isValidArea = area => includes( validAreas, area );
```
This helper function could then similarly be moved up. If there's a worry about confusing the function as a reducer, might be appropriate to split off a `utils.js` file. I always feel dirty doing it since it turns into something of a dumping ground, but at least helps alleviate this confusion.

aduth commented on client/state/ui/layout-focus/reducer.js:
```diff
@@ -0,0 +1,55 @@
+				return state;
+			}
+			// Temporary while replacing module with this reducer
+			layoutFocusModule.set( action.area );
```
I think in the case of `layoutFocus` it's not as big an issue since it's present everywhere anyways, but generally my worry with calling on modules from reducer (aside from side-effects no-no) is that reducers are loaded by boot and so are present in every section of the app. If this were a module more specific to a particular section, that would unnecessarily bloat the `commons` bundle.

sirbrillig replied:
Yeah, I don't love it, but it was the easiest way to keep parity during transition, and as you say, it's already everywhere 😝 

That said, happy to hear other ideas!

aduth commented on client/state/ui/layout-focus/reducer.js:
```diff
@@ -0,0 +1,55 @@
+			// Temporary while replacing module with this reducer
+			layoutFocusModule.next();
+			return Object.assign( {}, state, { current: state.next || 'content', previous: state.current, next: null } );
+		case SERIALIZE:
```
`state.ui` serialization is [disabled for all subtrees automatically](https://github.com/Automattic/wp-calypso/blob/3e9e2b7c9208bb6d8e7b576972dfd345f6eab16a/client/state/ui/reducer.js#L118-L124).

Further, I find the [`createReducer` helper](https://github.com/Automattic/wp-calypso/blob/3e9e2b7c9208bb6d8e7b576972dfd345f6eab16a/client/state/utils.js#L27-L94) (docs lacking, cc @gziolo 😄 ) to be quite nice to avoid having to manually specify these cases.

sirbrillig replied:
TIL

mtias replied:
This is an interesting one, it'd be nice to remember this between sessions if you close/open the app.

sirbrillig replied:
True, but maybe let's call that v2? 😁

aduth commented on client/state/ui/layout-focus/test/reducer.js:
```diff
@@ -0,0 +1,124 @@
+
+		it( 'does not set the current focus area if the value is invalid', function() {
+			const action = { type: LAYOUT_SET_FOCUS, area: 'foobar' };
+			const initialState = { current: 'content', previous: null, next: null };
```
I recommend using [`deepFreeze`](https://github.com/substack/deep-freeze) to ensure no accidental mutations of the original object.

sirbrillig replied:
Wise!

aduth commented on client/state/ui/layout-focus/test/selectors.js:
```diff
@@ -0,0 +1,44 @@
+import { getCurrentLayoutFocus, getPreviousLayoutFocus, getNextLayoutFocus } from '../../selectors';
+import layoutFocusModule from 'lib/layout-focus';
+
+let state;
```
Minor, but I'd move this `let` declaration into the `describe` block (so it's not processed unless the tests are actually run).

aduth commented on client/state/ui/selectors.js:
```diff
@@ -90,3 +91,39 @@ export function isPreviewShowing( state ) {
+ */
+export function getCurrentLayoutFocus( state ) {
+	// Temporary while we transition from the Flux store
+	return layoutFocusModule.getCurrent();
```
This is where I think it'd be nice if we were to have `layoutFocus` update Redux state so that we can always rely on our state, rather than call on `layoutFocus` from the reducer.

See [`lib/screen-title`](https://github.com/Automattic/wp-calypso/blob/master/client/lib/screen-title/index.js) as an example (looks like it was refactored recently into a middleware in #6669 by @yurynix).

sirbrillig replied:
I actually tried to do that first, but since it's available and used all kinds of places, we'd have to have access to the dispatcher and `getState` in the module. I couldn't come up with a way to do that. I guess #6669 came up with a way. :-) I may switch over to that, then.

sirbrillig replied:
After reviewing both versions of that change (#5208 and #6669), I'm still stuck with the same problem. Both of those cause the legacy action to be called when the redux store updates, but the problem here is that we want to update the redux store when the legacy action is called.

The only thing I can think of is to save a reference to the Redux store at some point just after it is created, then use that reference's dispatcher in the legacy Flux actions. I'll give that a try, but saving a reference to the store feels wrong somehow.

sirbrillig commented:
Updated so that the old `layoutFocus` module now sends its actions and queries to the Redux store instead. Seems to work well, actually, and is probably safer than the previous method.

aduth commented on client/lib/layout-focus/index.js:
```diff
@@ -1,9 +1,51 @@
+
+// Used to keep a reference to the Redux store so we can update it when the
+// legacy Flux store is updated.
+let reduxStore;
```
Do we need to think about SSR for this module? Setting the Redux store instance in a module variable is a bit worrisome if so. Or is that why there's fallbacks in the getter methods when store is not available?

sirbrillig replied:
I left the fallbacks for the situations when the Redux store hasn't yet been initialized, but this is a good point. Is the Redux store not available on the server? I don't remember the SSR documentation saying anything about that... but there's a lot I don't know about SSR here.

sirbrillig replied:
I did remove the DOM access if `window.document` is not available, which seemed to be the only SSR issue here.

ockham replied:
>  Is the Redux store not available on the server?

Yes it is, it's initialized [here](https://github.com/Automattic/wp-calypso/blob/be59246cf4e6b98c6c15f43161ce36357571051d/server/isomorphic-routing/index.js#L40).

ockham replied:
(and due to our isomorphic wrappers, it's available to router middleware from the `context` arg -- just like in a client-side setting)

sirbrillig replied:
@aduth I was thinking that we could remove the legacy layoutFocus code once we converted some of the early uses. 

On the other hand, I think it won't be used by anything, so instead we could just throw an error if something tries to use `layoutFocus` before it's bound to the Redux store now. That actually might be a quick way to find anything that might do that.

That said, I was hoping to make this PR relatively risk-free, and if we introduce a possible new error, that feels less safe to me. What do you think?

aduth replied:
My initial worry with binding Redux store this way is that on server side, since module references are shared, it _could_ open the possibility that a store instance is leaked between requests by different users.

sirbrillig replied:
Could we add a temporary check in here to see if it's on the server and, if so, just use the module state instead of Redux? (Temporary in that it would be removed when all uses are moved over to Redux and the module itself was removed.)

aduth commented on client/state/ui/layout-focus/reducer.js:
```diff
@@ -0,0 +1,45 @@
+			if ( ! isValidArea( action.area ) ) {
+				return state;
+			}
+			setFocusHideClass();
```
I'm trying to think of ways we can avoid this call being here. Could we create a Redux middleware for layoutFocus which watches for the `LAYOUT_FOCUS_SET` action?

sirbrillig replied:
I think that we could. Perhaps we could even investigate the necessity of that function at all and other work-arounds, but I felt that it was outside the scope of this PR.

aduth replied:
I took a closer look, and I don't think this class is even used anymore ([search](https://github.com/Automattic/wp-calypso/search?utf8=%E2%9C%93&q=focus-hide)). In fact, the original issue that this was trying to solve is now present once again, i.e. using browser search on sites section matches sites in Site picker even when site picker is not open (originally 2753-gh-calypso-pre-oss, cc @mtias). 

If we _were_ to reintroduce this, I wonder if we could do it by monitoring the `currentLayout` state from the sidebar component itself to add and remove the class.

In the meantime, I think we should just remove all traces of `setFocusHideClass`.

aduth commented on client/state/ui/selectors.js:
```diff
@@ -90,3 +90,33 @@ export function isPreviewShowing( state ) {
+ * @return {?String}  The current layout focus area
+ */
+export function getCurrentLayoutFocus( state ) {
+	return get( state, 'ui.layoutFocus.current' );
```
`state.uil.layoutFocus.current` will always be available, so you don't need `get`

sirbrillig replied:
Hm... it was last week so I'm having trouble remembering, but there was some situation where it wasn't available. I think maybe it's when these selectors are used by the boot code before the Redux store is initialized. This way they safely return null.

aduth replied:
>I think maybe it's when these selectors are used by the boot code before the Redux store is initialized.

I think we shouldn't be calling selectors with invalid state.

sirbrillig replied:
Fair point. I don't see that error any more, so who knows what I was doing. Removed.

sirbrillig commented:
Rebased to squash the old version away and fix conflicts.

sirbrillig commented:
Note to self: update layoutFocus README

aduth commented on client/state/ui/selectors.js:
```diff
@@ -98,3 +98,33 @@ export function getInitialQueryArguments( state ) {
+ * @param  {Object}  state Global state tree
+ * @return {?String}  The current layout focus area
+ */
+export function getCurrentLayoutFocus( state ) {
```
Should we move these selectors to `state/ui/layout-focus/selectors.js` ?

aduth replied:
>Should we move these selectors to `state/ui/layout-focus/selectors.js` ?

Same would go for `actions.js`

aduth commented:
Pushed 648e258 and 502e4fd to remove the `setFocusHideClass` references and try a middleware approach for syncing the legacy store with Redux. I like that it's a bit more of a separate unit that can be easily dropped once we've finished the transition. Let me know what you think.

aduth commented on client/state/ui/layout-focus/middleware.js:
```diff
@@ -0,0 +1,45 @@
+ * @return {Function} Redux middleware
+ */
+export default ( { dispatch, getState } ) => {
+	layoutFocus.on( 'change', () => {
```
Hmm, thinking again about SSR, I wonder if there are implications to registering an event listener for every Redux store instance that's initialized. Have you encountered this before @ockham ?

aduth replied:
There's a [couple of middlewares that are only registered in the `window` context](https://github.com/Automattic/wp-calypso/blob/e213bbcaf1f415f51d9c8466099cdcd0ba4fcb90/client/state/index.js#L82-L89). I intentionally chose not to include it there; the assumption being layout focus is not specific to client, though perhaps I'm wrong. In any case, we could include this middleware client-side only if need be if it's a temporary transition.

sirbrillig commented:
@aduth This is very clever. A combination of the two approaches I had tried. So when the module is used to change state, the middleware listener will update the redux state. When the actions are used to change state, the middleware will trigger a module update. Since both the module and redux states are idempotent, we avoid infinite loops. Previous state is kept in sync by the same mechanism.

The only thing that won't work here, I think, is if you do `layoutFocus.setNext('foo')` followed by `getNextLayoutFocus()` since the former won't actually update the redux store. However, the module doesn't have an accessor for getting the next state, so as long as nothing uses `getNextLayoutFocus`, we should be fine.

I'm still trying to wrap my mind around all the possible combinations of actions and accessors/selectors that could be run, but I can't immediately think of any other events that would break the sync, can you?

sirbrillig commented:
Actually, the reducer is not quite idempotent, right? The middleware keeps it from updating the same state twice. I wonder if we should add a similar check to the reducer itself, so that `setCurrentLayoutFocus('content'); setCurrentLayoutFocus('content');` won't change the state more than once?

sirbrillig commented:
Rebased to fix conflicts, moved actions and selectors to their own files, and made it so that `LAYOUT_FOCUS_SET` and `LAYOUT_NEXT_FOCUS_SET` called twice with the same action will only change the state once. I'm good to go with this if you are @aduth.

aduth commented on client/lib/layout-focus/index.js:
```diff
@@ -34,33 +49,25 @@ var layoutFocus = {
 	set: function( area ) {
-
-		if ( ! this.isValid( area ) || area === this._current ) {
+		if ( ! isValid( area ) || area === this._current ) {
```
Minor: Reversing the order here would allow the condition to be shortcut in the case that the area hasn't changed. Reference comparison [more performant](https://albertxing.com/stool/#734fd0b246ce057385a0400e0de7ee17) than `Array#indexOf` lookup in `isValid` and is probably the more likely reason for this function to be aborted:

```js
if ( area === this._current || ! isValid( area ) ) {
```

aduth commented:
@sirbrillig My only remaining concern is https://github.com/Automattic/wp-calypso/pull/6765/files#r71782170 . I do think this will accumulate event listeners infinitely on the server since they're never removed. I noted that we could observe this client-side only as a possible solution.

sirbrillig commented:
What if we only registered the listener once? 4c37b03 attempts to do that.

aduth commented:
>What if we only registered one listener at a time? 4c37b03 attempts to do that. 

No, that won't be enough unfortunately, since each request instantiates its own instance of the Redux store (one of the benefits of choosing Redux). With 4c37b03, some requests won't have the sync function triggered, since each store instance will run its own middleware separately. We might consider adding a clean-up action dispatched right before the response is output on the server, at which point we remove the listener.

[At this line perhaps?](https://github.com/Automattic/wp-calypso/blob/25d954b4fee06cedc66a0185349f352fad6b99ab/server/render/index.js#L93)

```js
context.store.dispatch( {
    type: SERVER_REQUEST_COMPLETE
} );
```

This is a pretty interesting problem for server-side rendering, and I'm now curious if there are other event listeners that are being attached without ever being cleaned up.

ockham commented:
> This is a pretty interesting problem for server-side rendering, and I'm now curious if there are other event listeners that are being attached without ever being cleaned up.

Curiously, we're running into something like that with a different PR right now: https://github.com/Automattic/wp-calypso/issues/6990#issuecomment-234944915

ockham commented:
Meh, ignore me. Should've read more context, sorry.

aduth commented:
@ockham Technically it's the same problem 😄  Something somewhere is attaching too many event handlers, possibly as a result of failure to clean up after it's "done".

sirbrillig commented:
Hm... tricky. What if we just removed all instances of the legacy `layoutFocus` module in this PR?

aduth commented on client/components/sidebar-navigation/index.jsx:
```diff
@@ -31,11 +32,12 @@ class SidebarNavigation extends React.Component {
 };
 
-export default SidebarNavigation;
+export default connect( null, { setLayoutFocus } )( SidebarNavigationSection );
```
I don't know that it'll be a problem, but keep in mind that `connect` makes components pure by default, which may or may not cause regressions if the component wasn't written well enough to be pure.

https://github.com/reactjs/react-redux/blob/master/docs/api.md#connectmapstatetoprops-mapdispatchtoprops-mergeprops-options

If there's a concern, you can pass `{ pure: false }` as the fourth argument to `connect`.

sirbrillig replied:
Oh, good point. I'll keep that in mind.

aduth commented:
>It's a ton of code changes, but it's starting to sound like it might be easier than trying to make a Flux store work with SSR.

Yeah, this might actually be the most reasonable approach, and has the benefit of expediting the transition 😄  I'd be fine to review it if this is the path you'd rather take.

aduth commented:
How many controllers aren't already wired to use `<ReduxProvider />`. That'd be something we'd need to include too, yeah? Maybe worth a separate PR upon which this could be rebased after?

sirbrillig commented:
> How many controllers aren't already wired to use <ReduxProvider />. That'd be something we'd need to include too, yeah? Maybe worth a separate PR upon which this could be rebased after?

Yeah, as I've been converting I was surprised how many there were. Definitely a separate PR, but I'm going to make the commits here for now and then `cherry-pick` them out.

sirbrillig commented:
There, that's all the uses of `layoutFocus` replaced with Redux versions. 

Not sure if I did `Loader` correctly, because I have no idea what that does.

Also clearly I've broken tests since I bet most of them don't wrap their components in Providers. I'll work on fixing those next.

sirbrillig commented:
That was easy. :-p

aduth commented:
@sirbrillig : What's the plan for this now? Did you want this to be reviewed and merged as-is, or were you going to pick away some smaller pull requests from the changes in this one?

sirbrillig commented:
@aduth I'm going to move the `Provider` commits into their own PR (#7198). I just added a commit here that that disables the layoutFocusMiddleware, so we should be able to test this PR to see if it works.

Testing will require trying to simulate all the layoutFocus shifts in each of these components. I've testing most of them myself, but I foolishly didn't write down which or how, so I'll start doing that now so I can get a list of all the ones I couldn't figure out how to test.

sirbrillig commented:
Places which use the new system:
- [ ] `splitTemplate` in server/bundler/loader: ???
- [x] `renderLayout` in boot/index: focus sidebar or sites if `sb` or `sp` is in the query string.
- [ ] `renderLayout` in boot/index: when code-splitting is enabled, move to the next layout focus (??)
- [x] `toggleSidebar` in `MobileBackToSidebar`: In the Reader, when clicking the header at mobile width (eg: "Followed Sites"), you should see the sidebar appear.
- [x] `toggleSidebar` in `SidebarNavigationSection`: in my-sites view, when clicking the header at mobile width (eg: "My WordPress Blog Stats"), you should see the sidebar appear.
- [x] `MasterBarLoggedIn`: clicking the my-sites button in the master bar from the reader should show the my-sites sidebar; clicking the reader button should hide the sidebar.
- [x] `DesignPreview`: should appear when the layout focus is set to `preview`; currently this is when clicking on the site logo in the my-sites sidebar. When closing the preview, the focus should return to the sidebar.
- [x] `Layout` css className should include `focus-${currentLayoutFocus}` that changes when the layout focus changes.
- [x] `onNavigate` in `MeSidebar`: clicking on any link in the sidebar at mobile width in the `/me` section should hide the sidebar.
- [ ] `sites` in my-sites/controller: sets the layout focus to content (??)
- [x] `switchSites` in `CurrentSite`: when clicking the "switch site" button in the my-sites header, the site picker should appear.
- [ ] `MySitesNavigation`: make sure `SitePicker` and `Sidebar` work below.
- [ ] `redirectToTeam` in my-sites/people/controller: navigate and then re-focus current layout focus called by `enforceSiteEnding` when no siteId is set. (?? is this still necessary with Redux?)
- [ ] `renderInvitePeople`in my-sites/people/controller: for jetpack sites, navigate and then re-focus current layout focus (?? is this still necessary with Redux?)
- [ ] `renderSingleTeamMember` in my-sites/people/controller: if the `PeopleLogStore` changes and has an error, navigate and then re-focus current layout focus (?? is this still necessary with Redux?)
- [x] `componentWillReceiveProps` in `SitePicker`: if the current layout focus is `sites` (i.e.: the site selector is showing), and we are not on a touch device (huh, weird), the `SiteSelector` component will get the prop `autoFocus` set to true.
- [x] `onClose` in `SitePicker`: if closing the site picker without the escape key, the sidebar should appear.
- [x] `closePicker` in `SitePicker`: if closing the site picker with the escape key, the sidebar should appear.
- [ ] `previewPost` in `PostActionsEllipsisMenuView`: clicking the `PopoverMenuItem` for "Preview" should show the post preview. (?? not sure how to find that button)
- [ ] `onNavigate` in `MySitesSidebar`: when clicking the "Stats" sidebar button at mobile width, the sidebar should disappear.
- [x] `onPreviewSite` in `MySitesSidebar`: when clicking the site icon in the sidebar, the preview should appear. 
- [ ] `focusContent` in `MySitesSidebar`: when clicking the "Add New Site" button in the sidebar at mobile width when you have only one site, the sidebar should disappear. 
- [ ] `redirectToDefaultSitePage` in my-sites/stats/controller: if a site is set when visiting a `/stats/...` route but the route does not exist, navigate to the default site's stats page and then re-focus current layout focus (?? is this still necessary with Redux?)
- [x] `toggleSidebar` in `EditorMobileNavigation`: when in the editor at mobile width, clicking the "Actions" button should show the sidebar.
- [x] `toggleSidebar` in `PostEditor`: when clicking on the "Write" button in the sidebar at mobile width, the sidebar should disappear.
- [x] `iframePreview` in `PostEditor`: when clicking "Preview" for a post in the editor at mobile width, the sidebar should disappear.
- [ ] Post previews are loading? This should be unrelated to these changes, but I can't get them to work. Could be an unrelated bug.
- [x] `handleClick` in `ReaderSidebar`: clicking on any element in the sidebar should cause the sidebar to disappear.
- [ ] `start` in signup/controller: after unmounting the secondary component, the layout focus should be set to 'content'. (??)

sirbrillig commented:
I may need help to test those unchecked items on the list above.

sirbrillig commented:
Rebased to fix conflicts. Will rebase again to squash if I can get all those test items checked off. Assistance appreciated!

aduth commented on client/components/mobile-back-to-sidebar/index.jsx:
```diff
@@ -28,4 +28,4 @@ var MobileBackToSidebar = React.createClass( {
 } );
 
-module.exports = MobileBackToSidebar;
+module.exports = connect( null, { setLayoutFocus } )( MobileBackToSidebar );
```
Minor: Consistency of imports/exports, could be `export default`

sirbrillig replied:
I just figured leaving `module.exports` was less risky since it works slightly differently than `export default` if someone uses `require`.

aduth replied:
>I just figured leaving `module.exports` was less risky since it works slightly differently than `export default` if someone uses `require`.

I think this is only true in modules including both named and default exports.

aduth commented on client/components/sidebar-navigation/index.jsx:
```diff
@@ -2,17 +2,23 @@
+class SidebarNavigationSection extends React.Component {
+	constructor( props ) {
+		super( props );
+		this.toggleSidebar = this.toggleSidebar.bind( this );
```
Was this broken previously? I could see how it might have been. Just curious about the changes here, and to the name of the component.

sirbrillig replied:
Oh, I guess the name might not be required. There are several components named `SidebarNavigation` and when testing I couldn't tell which one was this one, so I renamed it to be more specific.

I don't think it was technically broken because `toggleSidebar` didn't need to access `this` until this change, so the binding was not necessary.

aduth replied:
>I don't think it was technically broken because `toggleSidebar` didn't need to access `this` until this change, so the binding was not necessary.

Aha, this is an interesting exception to the rule that I probably don't consider as much as I ought to (i.e. not technically necessary to bind in all cases).

aduth commented on client/layout/index.jsx:
```diff
@@ -117,7 +119,9 @@ Layout = React.createClass( {
 				section={ this.props.section.group }
-				sites={ this.props.sites } />
+				sites={ this.props.sites }
+				setNextLayoutFocus={ this.props.setNextLayoutFocus }
```
Why not just `connect` `MasterbarLoggedIn` and allow it to call `setNextLayoutFocus` via its own `mapDispatchToProps` ?

sirbrillig replied:
A fair question. I was just trying to `connect` as few new components as possible with this change, since it can break tests. Do you think that would be a better flow here?

aduth replied:
>A fair question. I was just trying to `connect` as few new components as possible with this change, since it can break tests. Do you think that would be a better flow here?

I see the concern on avoiding breakage, but generally I'd say yes, we should be as granular as we need to be in directly connecting components to relevant state and dispatching props.

aduth commented on client/me/sidebar/index.jsx:
```diff
@@ -1,68 +1,72 @@
+
+const MeSidebar = React.createClass( {
 
 	displayName: 'MeSidebar',
```
`displayName` becomes redundant when `React.createClass` is assigned to a named variable.

aduth commented on client/my-sites/navigation/navigation.jsx:
```diff
@@ -2,46 +2,60 @@
 			<div className="sites-navigation">
 				<SitePicker
-					layoutFocus={ layoutFocus }
+					setNextLayoutFocus={ this.props.setNextLayoutFocus }
```
Again, ideally these components (`<SidePicker />` and `<Sidebar />`) can `connect` themselves to state.

sirbrillig replied:
Same as above: `connect`-ing components can break tests so I was just trying to keep things simple with this change. Do you think that would be a better flow?

aduth commented on client/post-editor/editor-mobile-navigation/index.jsx:
```diff
@@ -2,18 +2,19 @@
-const Site = require( 'blocks/site' ),
-	Gridicon = require( 'components/gridicon' ),
-	layoutFocus = require( 'lib/layout-focus' );
+const Site = require( 'blocks/site' );
```
See no reason these couldn't be `import`ed

sirbrillig replied:
Agreed. I was just trying to make as few changes as possible, but I guess I might as well here since they are already es6.

aduth commented on client/state/ui/layout-focus/reducer.js:
```diff
@@ -0,0 +1,42 @@
+			if ( action.area === state.current || ! isValidArea( action.area ) ) {
+				return state;
+			}
+			return Object.assign( {}, state, { current: action.area, previous: state.current } );
```
What do we ever use `previous` for? I smell YAGNI 😄

aduth commented on client/state/ui/layout-focus/reducer.js:
```diff
@@ -0,0 +1,42 @@
+			// previously, set it to `content`. This avoids having to set the
+			// focus to content on all navigation links because it becomes the
+			// default after focus has shifted.
+			if ( ! state.next && ! state.previous ) {
```
This condition and the fallback on the following line don't make sense to me. Why would lack of `previous` factor in to whether to return early? And if we remove that from the condition, then we don't need the fallback `|| 'content'` because it'll never be reached.

sirbrillig replied:
I don't fully understand the design (this behavior and comment were copied directly from the old module) but I *think* what this is for is when Calypso first loads, we automatically switch to `content` layout-focus. `previous` would thus be being used to determine if this is a new page load.

In any case, layoutFocus is so crucial to Calypso's functionality that I didn't want to risk breaking anything while we moved it over, which is why I left the functionality the same. Personally I'd rather improve the code by making additional PRs against the reducer rather than trying to do it all at once, but others may disagree.

aduth replied:
>I don't fully understand the design (this behavior and comment were copied directly from the old module)

Well, the nice thing about this pull request being all-inclusive now is that we see the full design, and I don't see anywhere we're touching `previous`. At least that's the basis for my comment.

sirbrillig replied:
Ok, I think I better understand the intent here:

On every page load, we automatically call `activateNextLayoutFocus`. If something has called `setNextLayoutFocus`, then shift `current` to that. However, if no there is no `next` value set, *and* this page load is originating from a sidebar or site picker, change `current` to `'content'`. This way, links which originate from the sidebar or site picker automatically hide those elements without having to do so explicitly.

`previous` is used to determine the part above where we need to know if the page load is originating from a sidebar or site picker.

That said, if the page load is *not* originating from the sidebar or site picker (or preview), it's likely originating from `content`, so there's really no danger in setting the current state to `content` in that case (or more precisely, setting the current state to `content` if there is no `next` and `content` is not the current state).

I'll give that a try.

sirbrillig replied:
Hm... I thought I grasped it, but I'm still not sure. I re-wrote that part of the reducer so that the logic is the same, but is now in the style of the original module. I guess what I'm not clear on is: when exactly is `activateNextLayoutFocus` called? It seems to be only in that weird `loader` file, but I don't understand when that happens in the normal process of navigation.

aduth commented on client/state/ui/layout-focus/test/actions.js:
```diff
@@ -0,0 +1,30 @@
+ * Internal dependencies
+ */
+import { LAYOUT_FOCUS_SET, LAYOUT_NEXT_FOCUS_SET, LAYOUT_NEXT_FOCUS_ACTIVATE } from 'state/action-types';
+import { setLayoutFocus, setNextLayoutFocus, activateNextLayoutFocus } from '../../layout-focus/actions';
```
`'../../layout-focus/actions'` could simply be `'../actions'`

aduth commented on server/bundler/loader.js:
```diff
@@ -84,7 +83,7 @@ function splitTemplate( path, section ) {
 		'	if ( _loadedSections[ ' + JSON.stringify( section.module ) + ' ] ) {',
 		'		controller.setSection( ' + JSON.stringify( section ) + ' )( context );',
-		'		layoutFocus.next();',
+		'		context.store.dispatch( { type: "LAYOUT_NEXT_FOCUS_ACTIVATE" } );',
```
We should probably `require` in the `activateNextLayoutFocus` action creator instead.

sirbrillig replied:
I tried that and it broke Calypso (see a few commits back). I don't even fully grasp how this works, but apparently we can't load action creators here?

aduth replied:
>I tried that and it broke Calypso (see a few commits back)

Looking at 05eb2dd02cacd3656cd8a8a5b3eb22d8704687f4, I think the issue is that you were trying to assign the action creator as the entire `actions` module, when it should have been:

```diff
-"\tactivateNextLayoutFocus = require( 'state/ui/layout-focus/actions' ),",
+"\tactivateNextLayoutFocus = require( 'state/ui/layout-focus/actions' ).activateNextLayoutFocus,",
```

sirbrillig replied:
oohhhhhh 😓 I was led astray because the other action creators in this block used raw constants and I thought maybe it was required. Red herring!

aduth replied:
>I was led astray

Yeah, noticed that too while going back through history. We should update those as well separately.

aduth commented on client/state/ui/layout-focus/reducer.js:
```diff
@@ -0,0 +1,42 @@
+export default function layoutFocus( state = initialState, action ) {
+	switch ( action.type ) {
+		case LAYOUT_FOCUS_SET:
+			if ( action.area === state.current || ! isValidArea( action.area ) ) {
```
Thinking on the `isValidArea`, should we be so proactive about validating? Thinking instead it's more valuable to fail loudly. Ideally (and probably for a subsequent pull request) we might extract these section names out into an object of constants.

sirbrillig replied:
> Thinking instead it's more valuable to fail loudly.

So, throw an error here? Or just allow any layoutfocus to be set and invalid ones will just fail to do anything/break stuff?

aduth replied:
>So, throw an error here? Or just allow any layoutfocus to be set and invalid ones will just fail to do anything/break stuff?

Don't think we need to throw, just don't like the idea of accommodating setting an invalid area (at least in that it's handled gracefully by this condition). Rather, I think we should technically allow for the invalid area to be assigned, insofar that it will obviously not work as expected and potentially cause unexpected behavior. But we should be diligent about ensuring that we're not doing that 😄

aduth commented on client/my-sites/navigation/navigation.jsx:
```diff
@@ -2,46 +2,60 @@
 					allSitesPath={ this.props.allSitesPath }
 					path={ this.props.path }
 					siteBasePath={ this.props.siteBasePath }
-					user={ this.props.user }
```
Missed a `this.props.user` reference in `<Sidebar />` causing an error:

https://github.com/Automattic/wp-calypso/blob/f3309ee/client/my-sites/sidebar/sidebar.jsx#L658

>sidebar.jsx:721 Uncaught TypeError: Cannot read property 'get' of undefined

sirbrillig replied:
Thank you for finding that!

aduth commented:
> previewPost in PostActionsEllipsisMenuView in PostTypeList: clicking the PopoverMenuItem for "Preview" should show the post preview. (?? not sure how to find that button)

Can't test because of the `<Sidebar>` error, but "View" or "Preview" should in the ellipsis menu should trigger an in-page preview on this screen:

http://calypso.localhost:3000/types/post

sirbrillig commented:
As an aside, I think I now understand all those calls that read `setNextLayoutFocus( currentLayoutFocus )`. They are to override the default behavior when switching routes which is to switch the focus back to `content`. Those areas must want to remain wherever they are.

sirbrillig commented:
Yup, removing `previous` from the state breaks the `?sb` and `?sp` query string behavior. I guess it does do something. :-/

sirbrillig commented:
For all the remaining unchecked test items above, I was able to test those sections of code, and they worked correctly. However, I was not able to trigger a condition where the layout focus code had any effect, which is why they aren't checked. That said, I believe they at least aren't doing any harm and I think this is in good shape.

aduth commented:
One thing I noticed is that previously, when modules called `layoutFocus.getCurrent()` and no focus had been assigned yet, it would return `content`, but now will return `null`. Is this a case we'll encountered and, if so, will it be an issue?

aduth commented:
Aside from the comment above, I couldn't find anything else needing attention. Took it for another spin and use-cases I tried all seem to work well.

sirbrillig commented:
> One thing I noticed is that previously, when modules called layoutFocus.getCurrent() and no focus had been assigned yet, it would return content

Good catch. I don't think that should have any effect, but I made a commit to change the default to 'content'. 

I'm going to squash this all now.

sirbrillig commented:
I think that's about as small as it's going to get and still have readable commits.
